src/foam/nanos/auth/AgentUserAuthService.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.ContextAwareSupport;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.nanos.NanoService;
+import foam.nanos.session.Session;
+
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.AND;
+
+public class AgentUserAuthService
+  extends    ContextAwareSupport
+  implements AgentAuthService, NanoService
+{
+  protected DAO userDAO_;
+  protected DAO groupDAO_;
+  protected DAO sessionDAO_;
+  protected DAO agentJunctionDAO_;
+
+  public AgentUserAuthService(X x) {
+    setX(x);
+  }
+
+  @Override
+  public void start() {
+    userDAO_     = (DAO) getX().get("localUserDAO");
+    groupDAO_    = (DAO) getX().get("groupDAO");
+    sessionDAO_  = (DAO) getX().get("sessionDAO");
+    agentJunctionDAO_  = (DAO) getX().get("agentJunctionDAO");
+  }
+
+  /* 
+    Sets the currently logged in user as an "agent" within the context &
+    sets "user" in the context to the passed in user. This allows users to
+    act on behalf of others while retaining information on the user.
+  */
+  public User ActAs(X x, User sudoUser) throws AuthenticationException {
+    User agent = (User) x.get("user");
+    User user = (User) userDAO_.find(sudoUser.getId());
+
+    if ( agent == null ) {
+      throw new AuthenticationException();
+    }
+
+    if ( user == null || ! user.getEnabled() ) {
```

Comment:
If we don't allow sudo-ing as disabled users then we won't be able to sudo as a Business because businesses will be disabled.

Reply:
I don't see why we'd limit a business user from logging in.

Reply:
I think there were two reasons:

1. That's what sudo-ing is for. The administrator of the business sudos into the business as opposed to them logging into the business directly.
2. The administrator's email address isn't tied to the business itself. That's good because if that administrator leaves, their email isn't tied to the login for the business.

Reply:
> The administrator's email address isn't tied to the business itself. That's good because if that administrator leaves, their email isn't tied to the login for the business.

Okay... If the administrator leaves and no one is left to be held accountable for the business than we'd have to implement some sort of claiming system or something along those lines. Which seems like added effort to me.

Plus this implementation isn't solely being created to satisfy our needs on the SME platform. It's a generic sudo system and I do think a user should be enabled to be logged into, considering that's how we flag deleted users in the system.

Reply:
The crux of the matter here is whether or not you should be able to log in as a business. That's what we need to come to an agreement on.

Here are the reasons I think we should not allow businesses to log in:

1. See above
2. See above
3. In Ablii our plan was to always require you to act as (sudo as) a business when you sign in. If you sign in as a business, what `User` are you going to act as?
4. Auditability. If you sign in as a business and modify a bunch of permissions, how do we know who was actually signed into the account? If we require you to log into your User account and act as the business before changing permissions we can track who's actually making the changes.

We need to figure this out because if we agree that you shouldn't be able to log in as a business, then this line would prevent you from sudo-ing as a business and we wouldn't be able to use this in Ablii, which is the whole reason we wanted it in the first place. If you still disagree and think you should be able to log in as a business, you need to tell me the solutions you have in mind for the 4 things I've outlined above and we can discuss from there. My guess is even if we could solve those problems, it would still be simpler to solve them by just removing part of the condition in this line, and the simplest solution is often the best one.

Reply:
> I do think a user should be enabled to be logged into, considering that's how we flag deleted users in the system.

Enabled is used to determine whether a `User` can log in, not to mark them as deleted. Also, we're not "logging in" to another user, we're acting as them.